M2/INSTALL-CMake.md

```diff
@@ -0,0 +1,327 @@
+
+**NOTE**: GCC or Clang compilers are recommended (in particular, not AppleClang). Optionally, install `ccache` for caching compiler artifacts and `ninja-build` (`ninja` on Brew) for optimized parallel builds.
+
+There are various tools needed to compile Macaulay2 dependencies.
+- On Debian/Ubuntu, install `autoconf build-essential bison libtool pkg-config yasm`.
+- On Mac OS X, using Homebrew, install `autoconf automake bison libtool pkg-config yasm`.
+
+There are 7 libraries that must be found on the system.
+- On Debian/Ubuntu, install `libopenblas-dev libeigen3-dev libxml2-dev libreadline-dev libgdbm-dev libboost-stacktrace-dev libatomic-ops-dev`.
+- On Mac OS X, using Homebrew, install `eigen libxml2 readline gdbm boost libatomic_ops`.
+
+Finally, there are 2 optional libraries that help with building other requirements.
+- On Debian/Ubuntu, install `libomp-dev libtbb-dev`.
+- On Mac OS X, using Homebrew, install `libomp tbb`.
+
+**NOTE**: the source directory must not contain any build artifacts from an in-source build. If you have built Macaulay2 in-source before, clean the build artifacts first by running `make clean distclean` in the source directory.
+
+A quick build involves the following steps:
+```
+git clone https://github.com/mahrud/M2.git -b feature/cmake
+cmake -GNinja -S M2/M2 -B M2/M2/BUILD/build -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr
+cmake --build M2/M2/BUILD/build --target build-libraries build-programs
+cmake --build M2/M2/BUILD/build --target install-packages
+cmake --build M2/M2/BUILD/build --target M2-emacs
+cmake --install M2/M2/BUILD/build
+```
+Each step is explained separately in the next section.
+
+### Building Macaulay2
+1. Clone Macaulay2 and switch to branch `feature/cmake`:
+```
+git clone https://github.com/mahrud/M2.git -b feature/cmake
+```
+
+2. Setup the build environment:
+```
+cd M2/M2/BUILD/build
+cmake -GNinja -S../.. -B. \
+      -DCMAKE_BUILD_TYPE=Release \
+      -DCMAKE_INSTALL_PREFIX=/usr
+```
+The `-S../..` argument indicates the location of the source and `-B.` indicates the build directory. After those, arguments of type `-DNAME=VALUE` set the `NAME` variable to `VALUE`. For instance, `CMAKE_BUILD_TYPE` determines various compiler flags to be used. Defined options are `Release`, `Debug`, `RelWithDebInfo`, and `RelMinSize`, with `RelWithDebInfo` being the default. The value of `CMAKE_INSTALL_PREFIX` determines the installation prefix.
+
+We build with MPIR as the multiple precision arithmetic library by default. To use GMP, use `-DUSING_MPIR=OFF` instead.
+
+This command generates the `build.ninja` files used by the Ninja build system, which is much more efficient. To generate a `Makefile` instead, remove `-GNinja` and use `make` instead of `ninja` in subsequent commands.
+
+3. Build the libraries that will be linked with the Macaulay2 executable:
+```
+ninja build-libraries
+```
+Note that this target **must** be built separately, before proceeding to `M2-binary`.
```

Comment:
Couldn't ninja be told to do this as a prerequisite for building M2-binary?

Reply:
It could, but that would be an anti-pattern. The idea is that ninja should **not** build libraries by default, because the default task is building M2. If a package isn't found, then the user should get an error.

The `build-libraries` target is a "[superbuild](https://blog.kitware.com/cmake-superbuilds-git-submodules/)" target, so it needs to come before the "build" step.

Reply:
This solves a bug with the comfigure/make system, which is that flint and a few other libraries are built every time you run make, and flint takes a long time to build, so you either need to know exactly which directories need to be rebuilt or wait a long time.

Reply:
Flint needing to be rebuilt is not a bug, it's a necessity -- there's no distribution out there yet that will work with our code.  What do you do about that?

Reply:
That's not what I meant. We also build Flint with `ninja build-flint`. But with configure/make, every time you run make Flint is recompiled, even if it is already compiled and exists in the same place.

Reply:
I think not.  Look at these files:
```
  /home/dan/src/M2/M2/BUILD/dan/builds.tmp/debian-development/libraries/flint:
  total used in directory 24 available 16117552
  drwxrwxr-x  3 dan dan  4096 May 23 09:22 .
  drwxrwxr-x 37 dan dan  4096 May 23 09:22 ..
  drwxrwxr-x 82 dan dan 12288 May 22 19:05 build
  -rw-rw-r--  1 dan dan     0 May 22 22:20 .checked-20200414
  -rw-rw-r--  1 dan dan     0 May 22 19:05 .compiled-20200414
  -rw-rw-r--  1 dan dan     0 May 22 18:51 .configured-20200414
  -rw-rw-r--  1 dan dan     0 May 22 19:05 .installed-20200414
  -r--r--r--  1 dan dan  1403 May 23 09:22 Makefile
  -rw-rw-r--  1 dan dan     0 May 22 18:51 .patched-20200414
```
We check for the presence of those various files of 0 length to decide whether something has been done already, so we don't do it again.

(That wrinkle hasn't been implemented yet for the submodules, which can be annoying.)

Reply:
Sure, but the `libraries` directory is very heavy, and I often run `make clean -C libraries` after the libraries and programs are installed in `usr-host`. Or I might want to make symlink from `usr-host` to the `usr-host` in another build directory in order to save time and space. In this case, configure doesn't detect that we've already built the right version of Flint. Forcing ninja to run `build-libraries` would cause it to rebuild the libraries, as the current build system does, even though they already exist in `usr-host`.

Reply:
Sorry I wasn't clear about "exists the same place" earlier. Shall I mark this conversation as resolved?

Reply:
Ah, I've never considered cleaning in `libraries` after building them.

And yes, flint offers no version number we can test for.

Resolved.

Reply:
One simple way to avoid rebuilding libraries and submodules is to run make in `Macaulay2/.`.  Or run `make all-in-Macaulay2` at top level.

Reply:
`flint/flint.h` includes the following:
```
#define FLINT_VERSION "2.6.0"
```
The standard practice with CMake is to use regex matching with header files to detect version, rather than compiling a program to extract version. Here is an example:
https://github.com/mahrud/M2/blob/feature/cmake/M2/cmake/FindFlint.cmake#L35-L46